### PR TITLE
Detect GitHub issue activity other than comments

### PR DESCRIPTION
**Pull request by JakeStoeffler**
##### Description
Implements: https://github.com/gitcoinco/web/issues/455

Uses the [GitHub timeline API](https://developer.github.com/v3/issues/timeline/) to detect various types of activity on a given issue performed by the interested party. Previously, only comments were considered. In this PR, the following events are considered "activity" and will prevent the inactivity warning email from being sent:

- `cross-referenced` - The issue was referenced from another issue
- `commented` - A comment was added to the issue
- `merge` - The issue was merged by the actor.
- `referenced` - The issue was referenced from a commit message.
- `review_requested` - review_requested

Also sets the `per_page` parameter to the max (100) for the GitHub comments and timeline APIs. This prevents comments from being missed (in most cases) if the API was returning multiple pages of results. HOWEVER, the issue of pagination still needs to be addressed in cases where there are more than 100 comments or timeline events. See [info about pagination in GitHub API](https://developer.github.com/v3/guides/traversing-with-pagination/). We might want to traverse pages of results in the future.

##### Checklist

- [x] linter status: 100% pass
- [ ] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)

marketing/management/commands

##### Testing

I tested locally by mocking some values into `expiration_start_work.py`. However, this should also be tested in a production-like environment before merging.

##### Refers/Fixes
Refs: https://github.com/gitcoinco/web/issues/455

### CODE REVIEW

**Comment by mkosowsk** [comment]
review_requested is very interesting to me and directly relates to the mentorship project my team is doing for the EthDenver hackathon currently 👍

**Comment by KennethAshley** [comment]
This came up as an [issue](https://gitcoin.co/funding/details?url=https://github.com/gitcoinco/web/pull/458) that needs to be worked on but it seems like this PR solves that issue?

**Comment by JakeStoeffler** [comment]
Thanks @mkosowsk and @KennethAshley. Yes, the original issue was here: https://gitcoin.co/funding/details?url=https://github.com/gitcoinco/web/issues/455

This PR implements the solution as well as some additional functionality to detect other types of activity.

**Comment by mkosowsk** [comment]
@JakeStoeffler awesome! I'm currently at EthDenver so won't be able to look at until probably Wednesday this week, but looking forward to it!

In the mean time, it looks like the CI build for travis has failed on #977. Could you please investigate to see if changes from this PR are the culprit?

Thanks!

**Comment by JakeStoeffler** [comment]
@mkosowsk @mbeacom Yes, I had fixed the isort issue in PR #459 but forgot to rebase my feature branch for this PR.  Travis build should pass now.

**Comment by mkosowsk** [comment]
Hey @JakeStoeffler seeing conflicts on expiration_start_work.py, could you please resolve? Thanks! 👍

**Comment by JakeStoeffler** [comment]
@mkosowsk Done, just needed another rebase.

**Comment by JakeStoeffler** [comment]
Lol, now the Travis build is failing, but the master branch build is also failing with the same error 😉

**Comment by mkosowsk** [comment]
Damn! 😂

Other than the major, major issue of copyright date being 2017 instead of 2018, this LGTM altho I haven't tested it 👍 

@mbeacom could you give this a once-over when you have a moment 👍 

Also, on Gitcoin it looks like this issue is "Open" to me so I can't sign off on the funding of it. @JakeStoeffler you may have to reclaim it?

https://gitcoin.co/funding/details?url=https://github.com/gitcoinco/web/pull/458

**Comment by JakeStoeffler** [comment]
@mkosowsk Hmm, there were two different Gitcoin issues and somehow they both got reset, so I reclaimed both 👍

**Comment by lazaridiscom** [comment]
@JakeStoeffler : sounds like double-or-nothing kind of thing...

**Comment by JakeStoeffler** [comment]
Wouldn't it be ironic if this PR fixes the very issue that caused both of those issues to reset? 😂 😂

**Comment by mkosowsk** [comment]
Like 10,000 spoons 😂

**Comment by owocki** [comment]
just catching up here.. 

code looks mostly good.  would like to test in a production-like environment before deploying... but currently busy with traiging a few other issues that came up when i was at ETHDenver.

if anyone wants to take the lead on testing this, you can import production data into your local with

```
./manage.py sync_geth mainnet 0 99999999999
```

**Comment by JakeStoeffler** [comment]
@owocki I will do that tonight. Didn't know about that trick :)

**Comment by JakeStoeffler** [comment]
@owocki Had issues getting that command to work. Doesn't seem to have pulled anything in. Ping me on Slack when you get a chance and we can figure out how to best test this PR.

**Comment by owocki** [comment]
just pinged u on slack

**Comment by JakeStoeffler** [comment]
@owocki I did some testing and discovered GitHub's timeline API doesn't always populate the `actor` field like [the docs](https://developer.github.com/v3/issues/timeline/) would make you believe.  Sometimes, they populate the `user` field instead, and for some events they don't populate either one. So, my code now handles that and I've tested it the best I can locally by generating some Interest objects and running `./manage.py expiration_start_work`.

**Comment by owocki** [comment]
awesome :)

**Comment by mkosowsk** [comment]
hey @JakeStoeffler, how's this guy coming along? 👍 

seeing another conflict on expiration_start_work.py, does this warrant another rebase?

@mbeacom and @owocki, if this conflict is resolved will the issue be ready to get paid out? deferring to your expertise and also don't want to keep jake in limbo 👍

**Comment by owocki** [comment]
sorry for the radio silence... @mbeacom and i were traveling during the flurry of activity above.. 

ill (or mark will) need to test this against some production data to confirm this is working.. @JakeStoeffler were you ever able to get the mainnet data synced?

**Comment by mkosowsk** [comment]
@owocki great, I will keep an eye on this 💃

**Comment by JakeStoeffler** [comment]
@owocki Yes I thought I mentioned it to you on Slack that I was able to get the data synced and as I stated above I was able to complete some testing using real data, but still not 100% production data.

@mkosowsk I am sure the conflict is due to needing a rebase.

Unfortunately I had a family emergency over the weekend and am OOTO for the next several days. Could someone take the lead here and see that this gets tested in prod and merged soon?

**Comment by owocki** [comment]
> Yes I thought I mentioned it to you on Slack that I was able to get the data synced 

you are right. i forgot about this.. my bad!

@JakeStoeffler shoot.. i am sorry to hear that!  @mbeacom and i need to do a review on this.  let us get some 👀 on testing this week

**Comment by owocki** [comment]
gotta resolve those merge conflicts

**Comment by owocki** [comment]
http://bits.owocki.com/423I2N3a100k/Screen%20Shot%202018-03-14%20at%202.02.39%20PM.png

i think the latest we should quote as a 'last heart from you time' is the 'start work' date

File: app/github/utils.py [hunk 1]
```diff
@@ -262,10 +263,21 @@ def search(query):
 
 
 def get_issue_comments(owner, repo, issue=None):
-    """Get the comments from issues on a respository."""
+    """Get the comments from issues on a respository.
+    PLEASE NOTE CURRENT LIMITATION OF 100 COMMENTS.
```
Comment (owocki) [inline app/github/utils.py]: would it be worth making multiple API calls here? and combining them?

File: app/github/utils.py [hunk 2]
```diff
@@ -275,6 +287,30 @@ def get_issue_comments(owner, repo, issue=None):
...
+        repo (str): Name of the repo
+        issue (int): Issue number
+
+    Returns:
+        requests.Response: The GitHub timeline response.
+    """
+    params = {
+        'sort': 'created',
+        'direction': 'desc',
+        'per_page': 100, #TODO traverse/concat pages: https://developer.github.com/v3/guides/traversing-with-pagination/
```
Comment (owocki) [inline app/github/utils.py]: same question as above

**Comment by owocki** [comment]
if i have a github bounty for issue X, which is commented on / mentioned on PR Y... does the timeline for X reflect the activity on PR Y ? i dont seem to get that gist from the github documentation... but maybe im wrong?

seems like we might have to go out and explicitly pull the events for anything related to PR Y in this case..

**Comment by owocki** [comment]
hey @JakeStoeffler are you still interested in this?  @mbeacom and i would like to bring it across teh finish line soon

**Comment by JakeStoeffler** [comment]
@owocki Yes, still interested -- just don't have a lot of free time atm. What all needs to be done on my end? Rebase and push? If you let me know I can possibly crank it out tonight.

**Comment by owocki** [comment]
rebase yes.. i left a few comments above WRT the actual scope too

**Comment by owocki** [comment]
thanks for doing that @mbeacom !

id certainly like to see the PR event pulls handled here.... is it a massive increase in scope or is it an easy thing to fold in?

**Comment by mbeacom** [comment]
@owocki It shouldn't be too bad.  I'll add it this evening.

**Comment by owocki** [comment]
coolio .  @ me back when ya want it tested

**Comment by mbeacom** [comment]
@owocki Where would you like to test this?

File: app/github/utils.py [hunk 3]
```diff
@@ -280,10 +281,21 @@ def search(query):
 
 
 def get_issue_comments(owner, repo, issue=None, comment_id=None):
-    """Get the comments from issues on a respository."""
+    """Get the comments from issues on a respository.
+    PLEASE NOTE CURRENT LIMITATION OF 100 COMMENTS.
```
Comment (owocki) [inline app/github/utils.py]: is it work traversing comments as part of the MVP? 

i dont think a lot of our open issues have > 100 comments.. but its still good to have the piece of mind of knowing that this works..

**Comment by owocki** [comment]
from testing locally:

> Exception in expiration_start_work.handle(): 'created_at'

**Comment by owocki** [comment]
another from testing locally: 

```
- its been 9 days since we heard from the user
executing should_delete_interest for mkosowsk / https://github.com/gitcoinco/web/issues/506
```

but mkosowsk has commented on this issue in the last 2 days

**Comment by owocki** [comment]
@mbeacom see my notes from testing above ^^

would be great to get this out this week.  we keep getting comments from ppl about how the gitcoinbot doesnt recognize the related PRs of an issue

**Comment by owocki** [comment]
> Exception in expiration_start_work.handle(): 'created_at'

@mbeacom is this still somethign youre going to add?  maybe we should ask someone in the community to steward this across the finish line?

**Comment by vs77bb** [comment]
@owocki @mbeacom FYI, this one is now in an 'Open' status on Gitcoin. Feel free to pub out to Slack (or lemme know if I should) 🙂

**Comment by vs77bb** [comment]
Hi @JakeStoeffler @mbeacom @owocki I think this one may have lost 'Start Work' due to inactivity. What needs to be done to take this over the finish line?

**Comment by JakeStoeffler** [comment]
@vs77bb Tbh I don't think I'm going to have the time to finish this up now.  Unfortunately I am swamped atm and this issue didn't get much feedback after I did the initial work on it.  Sorry!

**Comment by owocki** [comment]
just pushed up a few small commits to get this over the finish line.. i tested it.. it seems to be working now.. but the tests are failing

**Comment by owocki** [comment]
just pushed up a fix for the tests

**Comment by owocki** [comment]
@mkosowsk can you close the bounty here? https://gitcoin.co/issue/gitcoinco/web/458 

its been fixed already

**Comment by mkosowsk** [comment]
@owocki yep, just cancelled 👍

**Comment by weberV** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__This issue now has a funding of 0.05 ETH (34.61 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/gitcoinco/web/458/62).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $896.27 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by jagnarong** [comment]
Hi

**Pull request merged**